src/network/network.cpp

```diff
@@ -0,0 +1,74 @@
+#ifdef _WIN64 // WINDOWS
+#ifndef _MSC_VER
+#include <unistd.h>
+#endif
+#define _WINSOCK_DEPRECATED_NO_WARNINGS 1
+#ifndef WINSOCK2_IMPORTED
+#define WINSOCK2_IMPORTED
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#endif
+#include <windows.h>
+#else // NIX
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <stdio.h>
+#include <unistd.h>
+#endif // ...
+#include "system_state.hpp"
+#include "commands.hpp"
+#include "SPSCQueue.h"
+#include "network.hpp"
+
+namespace network {
+
+network_state::network_state() {
+	if(state.network_mode == sys::network_mode::single_player)
+		return; // Do nothing in singleplayer
+
+#ifdef _WIN64
+    WSADATA data;
+    if(WSAStartup(MAKEWORD(2, 2), &data) != 0)
+		std::abort();
+#endif
+	// TODO: Socket initialization here
+}
+
+void network_state::send_and_receive_commands(sys::state& state) {
+	if(state.network_mode == sys::network_mode::host) {
+		// put server commands onto the queues of the worker commands
+		auto* c = outgoing_commands.front();
+		while(c) {
+			// propagate to all the worker's queues
+			for(uint32_t i = 0; i < max_clients; i++)
+				if(clients[i].active)
+					clients[i].worker_commands.push(*c);
+			state.incoming_commands.push(*c); // and our local command queue
```

Comment:
this could result in a deadlock. If the incoming commands queue is full, this will wait. And if it is waiting, there is nothing running that could take them out of the queue. It also isn't clear how any of this will eventually work, since nothing ever gets taken out of worker commands. Are you going to block on each network call until it completes or are you going to use asynchronous io?

Reply:
asnchronous of course, threads will pop elements from the worker queues and the server, indeed, will block if there is a lot of commands - but that is practically unachievable unless you have 100's of players

Reply:
the server will not wait for the worker queues to empty, it will continue to push, the worker threads will pop at their own rate and this ensures that all commands that are about to be executed, are sent first into the clients and then put into the incoming commands queue

Reply:
You don't have to accept that possibility at all.Server can just receive command from client -> valid? -> execute and broadcast. It doesn't need to store it at all

Reply:
This is also where you would filter out console commands from clients

Reply:
broadcasting is blocking

Reply:
I thought that you were using async io?

Reply:
i was using raw sockets, should i've used asio??

Reply:
You don't have to block on a socket send. You can use a non-blocking send. Or, if your OS layer doesn't support that, dispatch the sending to a thread pool or something else that can block

Reply:
that is what i was doing, i was putting a thread for each client so it would block there with the workerqueue for the send/recv parts of the sockets - in fact i added a wrapper for non_blocking send and recv

Reply:
Also, you can create a non-blocking socket itself with
u_long mode = 1;  // 1 to enable non-blocking socket
ioctlsocket(sock, FIONBIO, &mode);

Reply:
that is what i was doing in fact lemme push the recv and send i was using for such

Reply:
Again, the exact implementation isn't important for the moment. My point is simply that by doing what I described above you don't need to then store the incoming commands, and thus you don't have to risk deadlocking